plugins/nf-prov/src/main/nextflow/prov/WrrocRenderer.groovy

```diff
@@ -0,0 +1,960 @@
+                        Path targetDir = dest.resolve(source.relativize(dir))
+                        Files.createDirectories(targetDir)
+                        return FileVisitResult.CONTINUE
+                    }
+
+                    @Override
+                    FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                        Path targetFile = dest.resolve(source.relativize(file))
+                        if (!Files.exists(targetFile))
+                            Files.copy(file, targetFile)
+                        return FileVisitResult.CONTINUE
+                    }
+                })
+            } else {
+                try {
+                    Files.createDirectories(dest.getParent())
+                    Files.copy(source, dest, StandardCopyOption.REPLACE_EXISTING)
+                } catch (Exception e) {
+                    println "workflowOutput Failed to copy $source to $dest: ${e.message}"
+                }
+            }
+        }
+
+        // get workflow config and store it in crate
+        final configFilePath = crateRootDir.resolve("nextflow.config")
+        final configFileWriter = new FileWriter(configFilePath.toString())
+        configMap.toConfigObject().writeTo(configFileWriter)
+
+        // get workflow README file and store it in crate
+        Map readmeFile = null
+
+        for( final fileName : README_FILENAMES ) {
+            final readmeFilePath = projectDir.resolve(fileName)
+            if( !Files.exists(readmeFilePath) )
+                continue
+
+            final encoding = FilenameUtils.getExtension(fileName).equals("md")
+                ? "text/markdown"
+                : "text/plain"
+            readmeFile = [
+                "@id"           : fileName,
+                "@type"         : "File",
+                "name"          : fileName,
+                "description"   : "This is the README file of the workflow.",
+                "encodingFormat": encoding
+            ]
+            Files.copy(readmeFilePath, crateRootDir.resolve(fileName), StandardCopyOption.REPLACE_EXISTING)
+            break
+        }
+
+        // get workflow metadata
+        final metadata = session.workflowMetadata
+        this.normalizer = new PathNormalizer(metadata)
+
+        final manifest = metadata.manifest
+        final nextflowMeta = metadata.nextflow
+        final scriptFile = metadata.getScriptFile()
+
+        final formatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME
+        final dateStarted = formatter.format(metadata.start)
+        final dateCompleted = formatter.format(metadata.complete)
+        final nextflowVersion = nextflowMeta.version.toString()
+        final wrrocParams = session.config.navigate('prov.formats.wrroc', [:]) as Map
+
+        // Copy workflow into crate directory
+        Files.copy(scriptFile, crateRootDir.resolve(scriptFile.getFileName()), StandardCopyOption.REPLACE_EXISTING)
```

Comment:
@simleo you mentioned copying all of the pipeline code into the crate, but is this really necessary? If the crate specifies a git repository, revision, and main script path, the user can reproduce the pipeline code at any time.

Reply:
> If the crate specifies a git repository, revision, and main script path

This would be a nice alternative, though including the code is safer since a git repository can be deleted.

Reply:
The use case I'm thinking of is using a standard pipeline e.g. nf-core/rnaseq v3.12, which may have been previously "verified" by a regulatory body. In this case it would be better to include an immutable reference rather than the actual files because then you don't have to check that those files correspond to a "verified" pipeline. But this is another topic I would like to see more discussion from community. I will consider a flag to copy the pipeline in a future iteration.

Reply:
@simleo Is there an appropriate entity for recording the git url + commit hash? For example the BCO has an "scm extension" that looks like this:

```json
"scm_extension": {
    "scm_repository": "https://github.com/nextflow-io/rnaseq-nf",
    "scm_type": "git",
    "scm_commit": "1815dc2a18bb2c2a8e4c7915260d77bb04ec8c91",
    "scm_path": "main.nf",
    "scm_preview": "https://github.com/nextflow-io/rnaseq-nf/tree/1815dc2a18bb2c2a8e4c7915260d77bb04ec8c91/main.nf"
}
```

Reply:
Add to the ComputationalWorkflow using codeRepository and version. Would be nice to also record the URL or path of the main script since it is the entrypoint to execute the workflow

Reply:
The ComputationalWorkflow could have a "url" property pointing to the exact revision, e.g.:

```
   "url": "https://github.com/famosab/wrrocmetatest/blob/e8665c295b12132aa886f3abd0fc34a71b08d4a6/main.nf"
```